File: swift/Sources/Ensoul/HTTPClient.swift
/// URLSession-based HTTP transport layer for the Ensoul Swift SDK.
///
/// Handles authentication, retries with exponential back-off, rate-limit
/// throttling, and SSE streaming — all through the standard URLSession API
/// (no external dependencies).
///
/// Auth types (`AuthProvider`, `APIKeyAuth`, `BearerAuth`) are defined in
/// `Auth.swift`. Rate-limit types (`RateLimitTracker`, `RateLimitInfo`) are
/// defined in `RateLimit.swift`.
import Foundation

// MARK: - SDK constants

private let sdkUserAgent = "ensoul-swift/0.1.0"
private let retryableStatusCodes: Set<Int> = [429, 500, 502, 503]

// MARK: - Retry helpers

/// Compute seconds to wait before the next retry.
///
/// Uses `Retry-After` when provided; otherwise falls back to exponential
/// back-off with jitter: `min(0.5 * 2^attempt, 30) + random(0, 1)`.
private func retryWait(attempt: Int, retryAfter: TimeInterval?) -> TimeInterval {
    if let ra = retryAfter, ra > 0 { return ra }
    let base = min(0.5 * pow(2.0, Double(attempt)), 30.0)
    let jitter = Double.random(in: 0..<1)
    return base + jitter
}

// MARK: - HTTPClient

/// URLSession-based HTTP client for the Ensoul API.
///
/// Usage:
/// ```swift
/// let config = ClientConfig(apiKey: "ens_...")
/// let client = HTTPClient(config: config)
///
/// // Simple GET
/// let (data, _) = try await client.get("/personas")
///
/// // Streaming SSE
/// for try await event in client.streamSSE(method: "POST", path: "/chat/stream", json: body) {
///     print(event.event, event.data)
/// }
/// ```
@available(iOS 15.0, macOS 12.0, *)
public class HTTPClient {
    private let config: ClientConfig
    private let auth: AuthProvider
    private let rateLimiter: RateLimitTracker
    private let session: URLSession
    /// Shared `JSONDecoder` (models use explicit CodingKeys for snake_case mapping).
    public let decoder: JSONDecoder

    // MARK: Init

    public init(config: ClientConfig, session: URLSession = .shared) {
        self.config = config
        self.session = session
        self.rateLimiter = RateLimitTracker()

        // Build auth strategy from config
        if let apiKey = config.apiKey {
            self.auth = APIKeyAuth(apiKey: apiKey)
        } else if let bearer = config.bearerToken {
            self.auth = BearerAuth(accessToken: bearer)
        } else {
            self.auth = APIKeyAuth(apiKey: "")
        }

        // Explicit CodingKeys on each model handle snake_case → camelCase mapping,
        // so no keyDecodingStrategy is needed (they would conflict).
        self.decoder = JSONDecoder()
    }

    // MARK: Path normalisation

    /// Ensure the path starts with `/v1/`.
    func normalizePath(_ path: String) -> String {
        let prefix = "/\(apiVersion)/"
        // Already versioned
        if path.hasPrefix(prefix) || path.hasPrefix("\(apiVersion)/") {
            return path.hasPrefix("/") ? path : "/\(path)"
        }
        let stripped = path.hasPrefix("/") ? String(path.dropFirst()) : path
        return "/\(apiVersion)/\(stripped)"
    }

    // MARK: URL building

    func buildURL(path: String, params: [String: String]?) throws -> URL {
        var base = config.baseURL
        while base.hasSuffix("/") { base = String(base.dropLast()) }
        let urlString = base + path

        guard var components = URLComponents(string: urlString) else {
            throw URLError(.badURL)
        }

        if let params, !params.isEmpty {
            components.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }
        }

        guard let url = components.url else {
            throw URLError(.badURL)
        }
        return url
    }

    // MARK: Default headers

    private func defaultHeaders(extra: [String: String]? = nil) -> [String: String] {
        var headers: [String: String] = [
            "User-Agent": sdkUserAgent,
            "Accept": "application/json",
        ]
        for (k, v) in config.customHeaders { headers[k] = v }
        for (k, v) in auth.authHeaders()   { headers[k] = v }
        if let extra { for (k, v) in extra { headers[k] = v } }
        return headers
    }

    // MARK: Core request method

    /// Make an authenticated HTTP request with retries and rate-limit handling.
    ///
    /// - Parameters:
    ///   - method:  HTTP method string (e.g. `"GET"`, `"POST"`).
    ///   - path:    API path — will be normalised to include `/v1/`.
    ///   - json:    Optional JSON body (encoded as UTF-8 JSON).
    ///   - params:  Optional URL query parameters.
    ///   - headers: Optional per-request headers merged on top of defaults.
    /// - Returns: Raw response `Data` and the `HTTPURLResponse`.
    public func request(
        method: String,
        path: String,
        json: [String: Any]? = nil,
        params: [String: String]? = nil,
        headers: [String: String]? = nil
    ) async throws -> (Data, HTTPURLResponse) {
        let normalizedPath = normalizePath(path)
        return try await _executeRequest(
            method: method,
            path: normalizedPath,
            json: json,
            params: params,
            headers: headers
        )
    }

    // MARK: Convenience methods

    public func get(
        _ path: String,
        params: [String: String]? = nil
    ) async throws -> (Data, HTTPURLResponse) {
        return try await request(method: "GET", path: path, params: params)
    }

    /// GET that skips `/v1/` prefix normalisation — for `/health` and similar
    /// endpoints that live outside the versioned API root.
    public func getRaw(
        _ path: String,
        params: [String: String]? = nil
    ) async throws -> (Data, HTTPURLResponse) {
        let url = try buildURL(path: path, params: params)
        var req = URLRequest(url: url, timeoutInterval: config.timeout)
        req.httpMethod = "GET"
        for (k, v) in defaultHeaders() { req.setValue(v, forHTTPHeaderField: k) }

        let (data, urlResponse) = try await session.data(for: req)
        guard let httpResponse = urlResponse as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        try raiseForStatus(data: data, response: httpResponse)
        return (data, httpResponse)
    }

    public func post(
        _ path: String,
        json: [String: Any]? = nil
    ) async throws -> (Data, HTTPURLResponse) {
        return try await request(method: "POST", path: path, json: json)
    }

    public func put(
        _ path: String,
        json: [String: Any]? = nil
    ) async throws -> (Data, HTTPURLResponse) {
        return try await request(method: "PUT", path: path, json: json)
    }

    public func patch(
        _ path: String,
        json: [String: Any]? = nil
    ) async throws -> (Data, HTTPURLResponse) {
        return try await request(method: "PATCH", path: path, json: json)
    }

    public func delete(_ path: String) async throws -> (Data, HTTPURLResponse) {
        return try await request(method: "DELETE", path: path)
    }

    /// POST with `application/x-www-form-urlencoded` body (used for OAuth
    /// token exchange endpoints).
    public func postForm(
        _ path: String,
        formData: [String: String]
    ) async throws -> (Data, HTTPURLResponse) {
        let normalizedPath = normalizePath(path)
        let url = try buildURL(path: normalizedPath, params: nil)

        var req = URLRequest(url: url, timeoutInterval: config.timeout)
        req.httpMethod = "POST"

        // Merge default headers, then override Content-Type
        for (k, v) in defaultHeaders() { req.setValue(v, forHTTPHeaderField: k) }
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        // Percent-encode each key=value pair and join with &
        let encoded = formData
            .map { k, v -> String in
                let ek = k.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? k
                let ev = v.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? v
                return "\(ek)=\(ev)"
            }
            .joined(separator: "&")
        req.httpBody = encoded.data(using: .utf8)

        let (data, urlResponse) = try await session.data(for: req)
        guard let httpResponse = urlResponse as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        rateLimiter.update(response: httpResponse)
        try raiseForStatus(data: data, response: httpResponse)
        return (data, httpResponse)
    }

    // MARK: SSE streaming

    /// Construct an SSE request and return an `SSEStream` that lazily initiates
    /// the connection when iterated.
    ///
    /// - Parameters:
    ///   - method: HTTP method (typically `"POST"` or `"GET"`).
    ///   - path:   API path — normalised to `/v1/` automatically.
    ///   - json:   Optional JSON request body.
    /// - Returns: An `SSEStream` `AsyncSequence` of `SSEEvent` values.
    public func streamSSE(
        method: String,
        path: String,
        json: [String: Any]? = nil
    ) -> SSEStream {
        let normalizedPath = normalizePath(path)

        // Build the URL (fall back to a placeholder on failure; iteration will
        // surface the error as a thrown value from next()).
        let url: URL
        if let u = try? buildURL(path: normalizedPath, params: nil) {
            url = u
        } else {
            // Force-unwrap safe: this literal always forms a valid URL.
            url = URL(string: "about:blank")!
        }

        var req = URLRequest(url: url, timeoutInterval: config.timeout)
        req.httpMethod = method

        var hdrs = defaultHeaders()
        hdrs["Accept"] = "text/event-stream"
        hdrs["Content-Type"] = "application/json"
        for (k, v) in hdrs { req.setValue(v, forHTTPHeaderField: k) }

        if let json {
            req.httpBody = try? JSONSerialization.data(withJSONObject: json)
        }

        return SSEStream(request: req, session: session)
    }

    // MARK: Private core request implementation

    private func _executeRequest(
        method: String,
        path: String,
        json: [String: Any]?,
        params: [String: String]?,
        headers: [String: String]?
    ) async throws -> (Data, HTTPURLResponse) {
        let url = try buildURL(path: path, params: params)

        var lastError: Error? = nil

        for attempt in 0...(config.maxRetries) {
            // Check rate-limiter before sending
            let result = rateLimiter.shouldWait()
            if result.wait {
                let nanos = UInt64(max(0, result.seconds) * 1_000_000_000)
                try await Task.sleep(nanoseconds: nanos)
            }

            // Build request
            var req = URLRequest(url: url, timeoutInterval: config.timeout)
            req.httpMethod = method

            for (k, v) in defaultHeaders(extra: headers) {
                req.setValue(v, forHTTPHeaderField: k)
            }

            if let json {
                req.setValue("application/json", forHTTPHeaderField: "Content-Type")
                req.httpBody = try JSONSerialization.data(withJSONObject: json)
            }

            do {
                let (data, urlResponse) = try await session.data(for: req)
                guard let httpResponse = urlResponse as? HTTPURLResponse else {
                    throw URLError(.badServerResponse)
                }

                // Update rate-limit state for future requests
                rateLimiter.update(response: httpResponse)

                let status = httpResponse.statusCode

                // Retryable responses — sleep and retry if we have attempts left
                if retryableStatusCodes.contains(status) && attempt < config.maxRetries {
                    var ra: TimeInterval? = nil
                    if status == 429,
                       let raw = httpResponse.value(forHTTPHeaderField: "Retry-After"),
                       let secs = TimeInterval(raw) {
                        ra = secs
                    }
                    let delay = retryWait(attempt: attempt, retryAfter: ra)
                    let nanos = UInt64(delay * 1_000_000_000)
                    try await Task.sleep(nanoseconds: nanos)
                    continue
                }

                // Non-retryable 4xx / 5xx — surface as SDK error
                try raiseForStatus(data: data, response: httpResponse)
                return (data, httpResponse)

            } catch let error as EnsoulAPIError {
                // Application-level errors (auth, validation, etc.) are never retried
                throw error
            } catch let error as URLError {
                lastError = error
                // Timeout or network errors — retry if we can
                if attempt < config.maxRetries {
                    let delay = retryWait(attempt: attempt, retryAfter: nil)
                    let nanos = UInt64(delay * 1_000_000_000)
                    try await Task.sleep(nanoseconds: nanos)
                    continue
                }
                throw error
            }
        }

        if let lastError { throw lastError }
        throw URLError(.unknown)
    }
}

File: cpp/src/http_client.cpp
#include "ensoul/http_client.hpp"
#include "ensoul/errors.hpp"
#include <httplib.h>
#include <thread>
#include <chrono>
#include <random>
#include <algorithm>
#include <set>
#include <sstream>
#include <cstdlib>

namespace ensoul {

static const std::set<int> RETRY_STATUS_CODES = {429, 500, 502, 503};
static const char* SDK_USER_AGENT = "ensoul-cpp/0.1.0";

std::string normalize_path(const std::string& path) {
    if (path.empty()) return "/v1/";
    if (path.find("/v1/") == 0 || path.find("/v1?") == 0) return path;
    if (path.find("/health") == 0 || path.find("/auth") == 0) return path;
    std::string p = path;
    if (p.front() != '/') p = "/" + p;
    return "/v1" + p;
}

static std::string build_query_string(const std::map<std::string, std::string>& params) {
    if (params.empty()) return "";
    std::ostringstream oss;
    bool first = true;
    for (const auto& [key, value] : params) {
        if (!first) oss << "&";
        first = false;
        // Simple URL encoding for common characters
        for (char c : key) {
            if (std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
                oss << c;
            } else {
                oss << '%' << std::uppercase << std::hex
                    << static_cast<int>(static_cast<unsigned char>(c));
            }
        }
        oss << '=';
        for (char c : value) {
            if (std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
                oss << c;
            } else {
                oss << '%' << std::uppercase << std::hex
                    << static_cast<int>(static_cast<unsigned char>(c));
            }
        }
    }
    return oss.str();
}

static double retry_wait(int attempt) {
    // Exponential backoff with jitter, capped at 30s
    double base = std::min(30.0, std::pow(2.0, attempt));
    static thread_local std::mt19937 gen(std::random_device{}());
    std::uniform_real_distribution<double> dist(0.0, base);
    return dist(gen);
}

struct HttpClient::Impl {
    httplib::Client client;
    ClientConfig config;
    std::unique_ptr<AuthProvider> auth;
    RateLimitTracker rate_limiter;

    Impl(const ClientConfig& cfg)
        : client(cfg.base_url), config(cfg) {
        // Set timeout (convert ms to seconds)
        int timeout_sec = static_cast<int>(cfg.timeout_ms / 1000);
        if (timeout_sec < 1) timeout_sec = 1;
        client.set_connection_timeout(timeout_sec);
        client.set_read_timeout(timeout_sec);
        client.set_write_timeout(timeout_sec);

        // Follow HTTP redirects (API uses 307 on paths without trailing slash)
        client.set_follow_location(true);

        // Build auth provider
        auth = make_auth(cfg.api_key, cfg.bearer_token);
    }
};

HttpClient::HttpClient(const ClientConfig& config)
    : impl_(std::make_unique<Impl>(config)) {}

HttpClient::~HttpClient() = default;

HttpResponse HttpClient::request(
    const std::string& method,
    const std::string& path,
    const nlohmann::json& json_body,
    const std::map<std::string, std::string>& params,
    const std::map<std::string, std::string>& extra_headers,
    bool stream)
{
    std::string normalized = normalize_path(path);

    // Append query params
    if (!params.empty()) {
        normalized += "?" + build_query_string(params);
    }

    // Build headers
    httplib::Headers headers;
    headers.emplace("User-Agent", SDK_USER_AGENT);
    headers.emplace("Accept", "application/json");

    // Auth headers
    if (impl_->auth) {
        for (const auto& [k, v] : impl_->auth->auth_headers()) {
            headers.emplace(k, v);
        }
    }

    // Custom headers from config
    for (const auto& [k, v] : impl_->config.custom_headers) {
        headers.emplace(k, v);
    }

    // Extra headers per-request
    for (const auto& [k, v] : extra_headers) {
        headers.emplace(k, v);
    }

    // If streaming, add Accept for SSE
    if (stream) {
        headers.erase("Accept");
        headers.emplace("Accept", "text/event-stream");
    }

    // Build body string
    std::string body_str;
    std::string content_type = "application/json";
    if (!json_body.is_null()) {
        body_str = json_body.dump();
    }

    int max_retries = impl_->config.max_retries;

    for (int attempt = 0; attempt <= max_retries; ++attempt) {
        // Check rate limit wait
        auto [should_wait, wait_time] = impl_->rate_limiter.should_wait();
        if (should_wait) {
            auto wait_ms = static_cast<long long>(wait_time * 1000);
            std::this_thread::sleep_for(std::chrono::milliseconds(wait_ms));
        }

        try {
            httplib::Result result;

            if (method == "GET") {
                result = impl_->client.Get(normalized, headers);
            } else if (method == "POST") {
                result = impl_->client.Post(normalized, headers, body_str, content_type);
            } else if (method == "PUT") {
                result = impl_->client.Put(normalized, headers, body_str, content_type);
            } else if (method == "PATCH") {
                result = impl_->client.Patch(normalized, headers, body_str, content_type);
            } else if (method == "DELETE") {
                result = impl_->client.Delete(normalized, headers);
            } else {
                throw EnsoulError("Unsupported HTTP method: " + method);
            }

            if (!result) {
                auto err = result.error();
                std::string err_msg = "HTTP request failed: ";
                switch (err) {
                    case httplib::Error::Connection:
                        err_msg += "Connection error"; break;
                    case httplib::Error::Read:
                        err_msg += "Read error"; break;
                    case httplib::Error::Write:
                        err_msg += "Write error"; break;
                    case httplib::Error::ConnectionTimeout:
                        err_msg += "Connection timeout"; break;
                    default:
                        err_msg += "Unknown error"; break;
                }

                if (attempt < max_retries) {
                    double wait = retry_wait(attempt);
                    std::this_thread::sleep_for(
                        std::chrono::milliseconds(static_cast<long long>(wait * 1000)));
                    continue;
                }
                throw EnsoulError(err_msg);
            }

            // Update rate limit tracker from response headers
            std::map<std::string, std::string> resp_headers;
            for (const auto& [k, v] : result->headers) {
                resp_headers[k] = v;
            }
            impl_->rate_limiter.update(resp_headers);

            int status = result->status;

            // Check for retryable status codes
            if (RETRY_STATUS_CODES.count(status) && attempt < max_retries) {
                double wait = retry_wait(attempt);
                // For 429, use Retry-After if available
                if (status == 429) {
                    auto it = resp_headers.find("Retry-After");
                    if (it == resp_headers.end()) it = resp_headers.find("retry-after");
                    if (it != resp_headers.end()) {
                        try {
                            wait = std::stod(it->second);
                        } catch (...) {}
                    }
                }
                std::this_thread::sleep_for(
                    std::chrono::milliseconds(static_cast<long long>(wait * 1000)));
                continue;
            }

            // Raise for non-stream error statuses
            if (!stream && status >= 400) {
                raise_for_status(status, result->body, resp_headers);
            }

            return HttpResponse{status, result->body, resp_headers};

        } catch (const EnsoulError&) {
            throw;
        } catch (const std::exception& e) {
            if (attempt < max_retries) {
                double wait = retry_wait(attempt);
                std::this_thread::sleep_for(
                    std::chrono::milliseconds(static_cast<long long>(wait * 1000)));
                continue;
            }
            throw EnsoulError(std::string("Request failed: ") + e.what());
        }
    }

    // Should not reach here, but just in case
    throw EnsoulError("Request failed after all retries");
}

HttpResponse HttpClient::post_form(
    const std::string& path,
    const std::map<std::string, std::string>& form_data)
{
    std::string normalized = normalize_path(path);

    httplib::Headers headers;
    headers.emplace("User-Agent", SDK_USER_AGENT);
    headers.emplace("Accept", "application/json");

    if (impl_->auth) {
        for (const auto& [k, v] : impl_->auth->auth_headers()) {
            headers.emplace(k, v);
        }
    }
    for (const auto& [k, v] : impl_->config.custom_headers) {
        headers.emplace(k, v);
    }

    // Build URL-encoded form body
    std::string body = build_query_string(form_data);
    std::string content_type = "application/x-www-form-urlencoded";

    auto result = impl_->client.Post(normalized, headers, body, content_type);

    if (!result) {
        throw EnsoulError("Form POST request failed: connection error");
    }

    std::map<std::string, std::string> resp_headers;
    for (const auto& [k, v] : result->headers) {
        resp_headers[k] = v;
    }
    impl_->rate_limiter.update(resp_headers);

    if (result->status >= 400) {
        raise_for_status(result->status, result->body, resp_headers);
    }

    return HttpResponse{result->status, result->body, resp_headers};
}

HttpResponse HttpClient::get_raw(
    const std::string& path,
    const std::map<std::string, std::string>& params)
{
    // No /v1/ prefix normalization for raw paths
    std::string full_path = path;
    if (!params.empty()) {
        full_path += "?" + build_query_string(params);
    }

    httplib::Headers headers;
    headers.emplace("User-Agent", SDK_USER_AGENT);
    headers.emplace("Accept", "application/json");

    if (impl_->auth) {
        for (const auto& [k, v] : impl_->auth->auth_headers()) {
            headers.emplace(k, v);
        }
    }
    for (const auto& [k, v] : impl_->config.custom_headers) {
        headers.emplace(k, v);
    }

    auto result = impl_->client.Get(full_path, headers);

    if (!result) {
        throw EnsoulError("GET raw request failed: connection error");
    }

    std::map<std::string, std::string> resp_headers;
    for (const auto& [k, v] : result->headers) {
        resp_headers[k] = v;
    }

    if (result->status >= 400) {
        raise_for_status(result->status, result->body, resp_headers);
    }

    return HttpResponse{result->status, result->body, resp_headers};
}

std::string HttpClient::stream_sse_raw(
    const std::string& method,
    const std::string& path,
    const nlohmann::json& json_body,
    const std::map<std::string, std::string>& params)
{
    // Use the request method with stream=true to get the raw SSE body
    auto resp = request(method, path, json_body, params, {}, true);

    if (resp.status_code >= 400) {
        raise_for_status(resp.status_code, resp.body, resp.headers);
    }

    return resp.body;
}

// Convenience methods
HttpResponse HttpClient::get(const std::string& path,
                              const std::map<std::string, std::string>& params) {
    return request("GET", path, nullptr, params);
}

HttpResponse HttpClient::post(const std::string& path,
                               const nlohmann::json& json_body,
                               const std::map<std::string, std::string>& params) {
    return request("POST", path, json_body, params);
}

HttpResponse HttpClient::put(const std::string& path,
                              const nlohmann::json& json_body) {
    return request("PUT", path, json_body);
}

HttpResponse HttpClient::patch(const std::string& path,
                                const nlohmann::json& json_body) {
    return request("PATCH", path, json_body);
}

HttpResponse HttpClient::del(const std::string& path) {
    return request("DELETE", path);
}

} // namespace ensoul
